src/azure-cli-core/azure/cli/core/adal_authentication.py

```diff
@@ -77,17 +75,56 @@ def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
 
         _, token, full_token, _ = self._get_token(_try_scopes_to_resource(scopes))
 
-        try:
-            expires_on = full_token['expiresOn']
-            return AccessToken(token, int(datetime.datetime.strptime(expires_on, '%Y-%m-%d %H:%M:%S.%f').timestamp()))
-        except:  # pylint: disable=bare-except
-            pass  # To avoid crashes due to some unexpected token formats
-
-        try:
-            return AccessToken(token, int(full_token['expiresIn'] + time.time()))
-        except KeyError:  # needed to deal with differing unserialized MSI token payload
+        # NEVER use expiresIn (expires_in) as the token is cached and expiresIn will be already out-of date
+        # when being retrieved.
+
+        # User token entry sample:
+        # {
+        #     "tokenType": "Bearer",
+        #     "expiresOn": "2020-11-13 14:44:42.492318",
+        #     "resource": "https://management.core.windows.net/",
+        #     "userId": "test@azuresdkteam.onmicrosoft.com",
+        #     "accessToken": "eyJ0eXAiOiJKV...",
+        #     "refreshToken": "0.ATcAImuCVN...",
+        #     "_clientId": "04b07795-8ddb-461a-bbee-02f9e1bf7b46",
+        #     "_authority": "https://login.microsoftonline.com/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a",
+        #     "isMRRT": True,
+        #     "expiresIn": 3599
+        # }
+
+        # Service Principal token entry sample:
+        # {
+        #     "tokenType": "Bearer",
+        #     "expiresIn": 3599,
+        #     "expiresOn": "2020-11-12 13:50:47.114324",
+        #     "resource": "https://management.core.windows.net/",
+        #     "accessToken": "eyJ0eXAiOiJKV...",
+        #     "isMRRT": True,
+        #     "_clientId": "22800c35-46c2-4210-b8a7-d8c3ec3b526f",
+        #     "_authority": "https://login.microsoftonline.com/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a"
+        # }
+        if 'expiresOn' in full_token:
+            import datetime
+            expires_on_timestamp = int(_timestamp(
+                datetime.datetime.strptime(full_token['expiresOn'], '%Y-%m-%d %H:%M:%S.%f')))
+            return AccessToken(token, expires_on_timestamp)
+
+        # Cloud Shell (Managed Identity) token entry sample:
+        # {
+        #     "access_token": "eyJ0eXAiOiJKV...",
+        #     "refresh_token": "",
+        #     "expires_in": "2106",
+        #     "expires_on": "1605686811",
+        #     "not_before": "1605682911",
+        #     "resource": "https://management.core.windows.net/",
+        #     "token_type": "Bearer"
+        # }
+        if 'expires_on' in full_token:
             return AccessToken(token, int(full_token['expires_on']))
 
+        from azure.cli.core.azclierror import CLIInternalError
+        raise CLIInternalError("No expiresOn or expires_on is available in the token entry.")
```

Comment:
If there is no expected property in the token entry, should it be AAD or ADAL issue?

Reply:
As I tested, it is never possible for `expiresOn` or `expires_on` to be missing.